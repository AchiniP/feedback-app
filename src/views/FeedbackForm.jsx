import React from 'react';
import _ from 'lodash';
import { withStyles, makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import {
  Button,
  Dialog,
  DialogTitle as MuiDialogTitle,
  DialogContent as MuiDialogContent,
  DialogActions as MuiDialogActions,
  IconButton,
  Typography,
  TextField,
} from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
// eslint-disable-next-line import/no-cycle
import ThankYouCard from './ThankYouCard';
import FormHook from '../logic/FormHook';
import isValidEmail from '../logic/Validation';

const useStyles = makeStyles(theme => ({
  root: {
    margin: 0,
    padding: theme.spacing(2),
    background: 'black',
    color: 'white',
    textAlign: 'center',
  },
  container: {
    background: 'lightBlue',
  },
  dialogPaper: {
    maxHeight: '80vh',
    maxWidth: '50%',
    justifyContent: 'center',
    position: 'absolute',
    bottom: '0',
    left: '0',
  },
  closeButton: {
    position: 'absolute',
    right: theme.spacing(1),
    top: theme.spacing(1),
    color: 'white',
  },
}));

const DialogTitle = ((props) => {
  const classes = useStyles();
  const {
    children, onClose, ...other
  } = props;
  return (
    <MuiDialogTitle disableTypography className={classes.root} {...other}>
      <Typography variant="h6">{children}</Typography>
      {onClose ? (
        <IconButton
          aria-label="close"
          className={classes.closeButton}
          onClick={onClose}
        >
          <CloseIcon />
        </IconButton>
      ) : null}
    </MuiDialogTitle>
  );
});

const DialogContent = withStyles(theme => ({
  root: {
    padding: theme.spacing(2),
  },
}))(MuiDialogContent);

const DialogActions = withStyles(theme => ({
  root: {
    margin: 0,
    padding: theme.spacing(1),
  },
}))(MuiDialogActions);

export default function FeedbackForm(props) {
  const { formOpen, onClose } = props;
  const classes = useStyles();
  const [open, setOpen] = React.useState(false);

  const handleThankyouCardOpen = () => {
    setOpen(true);
  };

  const { inputs, handleInputChange, handleSubmit } = FormHook({ like: '', dislike: '', email: '' }, handleThankyouCardOpen);
  return (
    open ? <ThankYouCard onClose={onClose} content="Thank you!" subContent="Your feedback is valueable to us." />
      : (
        <Dialog
          classes={{ paper: classes.dialogPaper, container: classes.container }}
          onClose={onClose}
          aria-labelledby="customized-dialog-title"
          open={formOpen}
          fullWidth
        >
          <DialogTitle id="customized-dialog-title" onClose={onClose}>
            Tell us more
          </DialogTitle>
          <DialogContent dividers>
            <TextField
              id="like"
              name="like"
              type="text"
              label="What did you like the most?"
              style={{ margin: 8 }}
              placeholder="Tell us your experience (optional)"
              fullWidth
              margin="normal"
              InputLabelProps={{
                shrink: true,
              }}
              onChange={handleInputChange}
              value={inputs.like}
            />
            <TextField
              id="dislike"
              name="dislike"
              type="text"
              label="What did you like least?"
              style={{ margin: 8 }}
              placeholder="Let us know how we can improve (optional)"
              fullWidth
              margin="normal"
              InputLabelProps={{
                shrink: true,
              }}
              onChange={handleInputChange}
              value={inputs.dislike}
            />
            <TextField
              id="email"
              name="email"
              type="email"
              label="Your email"
              style={{ margin: 8 }}
              placeholder="Your email address (optional)"
              fullWidth
              margin="normal"
              InputLabelProps={{
                shrink: true,
              }}
              onChange={handleInputChange}
              value={inputs.email}
              error={!isValidEmail(inputs.email)}
              helperText={isValidEmail(inputs.email) ? '' : 'Please enter a valid email address.'}
            />

          </DialogContent>
          <DialogActions>
            <Button
              id="submit-button"
              autoFocus
              color="primary"
              onClick={handleSubmit}
              disabled={_.isEmpty(inputs.like) && _.isEmpty(inputs.dislike) && (_.isEmpty(inputs.email) || !isValidEmail(inputs.email))}
            >
              SUBMIT
            </Button>
          </DialogActions>
        </Dialog>
      )
  );
}

FeedbackForm.propTypes = {
  onClose: PropTypes.func.isRequired,
  formOpen: PropTypes.bool.isRequired,
};

DialogTitle.propTypes = {
  onClose: PropTypes.func.isRequired,
  children: PropTypes.string,
};
