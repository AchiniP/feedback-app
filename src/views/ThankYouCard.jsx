import React from 'react';
import PropTypes from 'prop-types';
import { 
  makeStyles, Card, CardContent, Typography, ClickAwayListener, Grid 
} from '@material-ui/core';
import SentimentSatisfiedTwoToneIcon from '@material-ui/icons/SentimentSatisfiedTwoTone';
// eslint-disable-next-line import/no-cycle
import FeedbackForm from './FeedbackForm';

const useStyles = makeStyles({
  root: {
    width: '70%',
    // height: '20%',
    textAlign: 'center',
    position: 'absolute',
    bottom: 5,
    // padding: '5%',
  },
  icon: {
    verticalAlign: 'middle',
    fill: '#d6b911',
  },
});

export default function ThankYouCard(props) {
  const classes = useStyles();
  const {
    onClose, showIcon, content, subContent,
  } = props;
  const [formOpen, setFormOpen] = React.useState(false);
  const [hideCard, setHideCard] = React.useState(false);

  const handleClickOpen = () => {
    setFormOpen(showIcon);
    setHideCard(!showIcon);
  };

  return (
    // eslint-disable-next-line no-nested-ternary
    formOpen ? <FeedbackForm formOpen={formOpen} onClose={onClose} />
      : (hideCard ? ''
        : (
          <ClickAwayListener onClickAway={handleClickOpen}>
            <Grid container>
              <Grid item xs />
              <Grid item xs={8}>
                <Card
                  className={classes.root}
                  maxwidth="lg"
                >
                  <CardContent>
                    <Typography variant="h6" color="textSecondary" component="p">
                      {showIcon ? <SentimentSatisfiedTwoToneIcon className={classes.icon} /> : ''}
                      {' '}
                      { content }
                    </Typography>
                    {subContent
                      ? (
                        <Typography variant="body2" color="textSecondary" component="p">
                          {showIcon ? <SentimentSatisfiedTwoToneIcon className={classes.icon} /> : ''}
                          {' '}
                          { subContent }
                        </Typography>
                      )
                      : ''}
                  </CardContent>
                </Card>
              </Grid>
              <Grid item xs />
            </Grid>

          </ClickAwayListener>
        )
      )
  );
}

ThankYouCard.propTypes = {
  onClose: PropTypes.func.isRequired,
  showIcon: PropTypes.bool,
  content: PropTypes.string.isRequired,
  subContent: PropTypes.string,
};