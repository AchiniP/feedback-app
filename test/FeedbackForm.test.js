import React from 'react';
import ReactDOM from 'react-dom';
import { mount, shallow, render } from 'enzyme';
import renderer from 'react-test-renderer';
import FeedbackForm from '../src/views/FeedbackForm';
import ThankYouCard from '../src/views/ThankYouCard';

describe('Should Render FeedbackForm without Crashing', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    const onCLoseFn = jest.fn;
    ReactDOM.render(<FeedbackForm formOpen onClose={onCLoseFn} />, div);
  });
});

describe('Snapshot testing FeedbackForm', () => {
  const component = mount(
    <FeedbackForm formOpen onClose={() => {}} />,
  );
  expect(component.html()).toMatchSnapshot();
});

describe('Test clicks', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <FeedbackForm formOpen onClose={() => {}} />,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should update input fields', () => {
    const event = {
      preventDefault() {},
      target: { value: 'test', name: 'email' },
    };
    const email = wrapper.find('input[id="email"]');
    expect(email.exists()).toEqual(true);
    email.simulate('change', event);
    const updatedWrapper = wrapper.update();
    const updatedEmail = updatedWrapper.find('input[id="email"]');
    expect(updatedEmail.props().value).toEqual('test');
  });

  it('should simulate button click', () => {
    const event = {
      preventDefault() {},
    };
    const submitButton = wrapper.find('button[id="submit-button"]');
    expect(submitButton.exists()).toEqual(true);
    submitButton.simulate('click', event);
    const thanksCard = wrapper.find(ThankYouCard);
    expect(thanksCard.exists()).toEqual(true);
  });

  it('should call Thank you card component with correct props', () => {
    const thanksCard = wrapper.find(ThankYouCard);
    expect(thanksCard.props().content).toEqual('Thank you!');
    expect(thanksCard.props().subContent).toEqual('Your feedback is valueable to us.');
  });
});
