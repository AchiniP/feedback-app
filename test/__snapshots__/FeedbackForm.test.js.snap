// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"<div role=\\"presentation\\" class=\\"MuiDialog-root\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\"><div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div><div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div><div class=\\"MuiDialog-container makeStyles-container-2 MuiDialog-scrollPaper\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"none presentation\\" tabindex=\\"-1\\"><div class=\\"MuiPaper-root MuiDialog-paper makeStyles-dialogPaper-3 MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded\\" role=\\"dialog\\" aria-labelledby=\\"customized-dialog-title\\"><div class=\\"MuiDialogTitle-root makeStyles-root-1\\" id=\\"customized-dialog-title\\"><h6 class=\\"MuiTypography-root MuiTypography-h6\\">Tell us more</h6><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-closeButton-4\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"close\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div><div class=\\"MuiDialogContent-root WithStyles(ForwardRef(DialogContent))-root-102 MuiDialogContent-dividers\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\" style=\\"margin: 8px;\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink\\" data-shrink=\\"true\\" for=\\"like\\" id=\\"like-label\\">What did you like the most?</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"like\\" name=\\"like\\" placeholder=\\"Tell us your experience (optional)\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\" style=\\"margin: 8px;\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink\\" data-shrink=\\"true\\" for=\\"dislike\\" id=\\"dislike-label\\">What did you like least?</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"dislike\\" name=\\"dislike\\" placeholder=\\"Let us know how we can improve (optional)\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\" style=\\"margin: 8px;\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink\\" data-shrink=\\"true\\" for=\\"email\\" id=\\"email-label\\">Your email</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"email\\" name=\\"email\\" placeholder=\\"Your email address (optional)\\" type=\\"email\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiDialogActions-root WithStyles(ForwardRef(DialogActions))-root-162 MuiDialogActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" id=\\"submit-button\\"><span class=\\"MuiButton-label\\">SUBMIT</span></button></div></div></div><div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div></div>"`;
